### PR TITLE
Kcho/upenn redcap

### PR DESCRIPTION
**Pull request by kcho**
UPENN neuropsych REDCap source has been set up in a way to use "session_subid" to have the AMP-SCZ ID. (Note that this REDCap is for neuropsychological tests only, which is different from the general REDCap that we are using for AMP-SCZ project)

This PR is tested on Pronet data aggregation server, and fully working.

### CODE REVIEW

File: lochness/redcap/__init__.py [hunk 1]
```diff
@@ -333,12 +333,23 @@ def sync(Lochness, subject, dry=False):
...
-                'records': redcap_subject
-            }
+            if 'UPENN' in redcap_instance:
+                # UPENN REDCap is set up with its own record_id, but have added
+                # "session_subid" field to note AMP-SCZ ID
+                record_query = {
+                    'token': api_key,
+                    'content': 'record',
+                    'format': 'json',
+                    'filterLogic': f"[session_subid] = '{redcap_subject}'"
```
Comment (tashrifbillah) [inline lochness/redcap/__init__.py:343]: Shouldn't the last element be `subject.id`?

Comment (tashrifbillah) [reply lochness/redcap/__init__.py:343]: Second question:

> r = requests.post(api_url, data=data, stream=True, verify=False)

Does a REDCap query recognize `filterLogic`?

**Comment by kcho** [comment]
@tashrifbillah Added `or` logic to REDCap API to accommodate UPENN and Sylvain's suggestion of CAPS agonistic ID search. fbeade8

File: lochness/redcap/__init__.py (hunk 1 above)
Comment (kcho) [reply lochness/redcap/__init__.py]: Here `redcap_subject` is also the ID. REDCap query recognizes `filterLogic`. I've check this PR in Pronet server with YA00015 and it's fully working.

File: lochness/redcap/__init__.py [hunk 2]
```diff
@@ -333,12 +333,25 @@ def sync(Lochness, subject, dry=False):
...
+            if 'UPENN' in redcap_instance:
+                # UPENN REDCap is set up with its own record_id, but have added
+                # "session_subid" field to note AMP-SCZ ID
+                redcap_subject_sl = redcap_subject.lower()
+                record_query = {
+                    'token': api_key,
+                    'content': 'record',
+                    'format': 'json',
+                    'filterLogic': f"[session_subid] = '{redcap_subject}' or "
+                                   f"[session_subid] = '{redcap_subject_sl}'"
```
Comment (tashrifbillah) [inline lochness/redcap/__init__.py:345]: Does this assignment work in Python? :o

Comment (kcho) [reply lochness/redcap/__init__.py:345]: Yes, it works. When the line breaks are within `{}`, `()`, or `[]`, you do not have to add `\` at the end of the string to concatenate them.

Comment (tashrifbillah) [reply lochness/redcap/__init__.py:345]: Right I meant does this complex query work for REDCap?

Comment (kcho) [reply lochness/redcap/__init__.py:345]: Yes. It does.

**Comment by tashrifbillah** [comment]
< to accommodate UPENN and Sylvain's suggestion of CAPS agnostic ID search.

Why don't we do only lower case search then? Can REDCap not return the intended ID regardless of case?

**Comment by kcho** [comment]
> Why don't we do only lower case search then? Can REDCap not return the intended ID regardless of case?

Pronet team has been instructing other sites to use CAPs letter only, but our thought was that some site may enter them in lower cases. This additional `or` logic is to capture this mistake.

**Comment by tashrifbillah** [comment]
Let's consider this example: REDCap has ID `ABC`. If I write `abc` in `filterLogic`, can it retrieve 'ABC'?

**Comment by kcho** [comment]
>Let's consider this example: REDCap has ID ABC. If I write abc in filterLogic, can it retrieve ABC?

No. It cannot.

**Review by tashrifbillah — APPROVED** [review]

**Pull request merged**